File: jira_db_sync/jira_sync/sync.py
"""同期サービス

責務:
- 全件同期 / 差分同期 / レジューム同期の制御
- チェックポイント管理（sync_state.json）
- 同期フロー全体のオーケストレーション
"""

import json
import logging
import os
from datetime import datetime, timedelta, timezone
from typing import Callable

import pandas as pd

from jira_sync.client import JiraClient
from jira_sync.db import Database
from jira_sync.fields import FieldExpander
from jira_sync.transform import ChangeHistoryTransformer, IssueTransformer

logger = logging.getLogger(__name__)

MARGIN_MINUTES = 5  # 差分同期のマージン（JQL の分単位精度を補完）


class SyncState:
    """同期状態の永続化 (JSON ファイル)"""

    def __init__(self, path: str):
        self.path = path

    def _load(self) -> dict:
        if os.path.exists(self.path):
            try:
                with open(self.path) as f:
                    return json.load(f)
            except Exception:
                pass
        return {}

    def _save(self, state: dict):
        os.makedirs(os.path.dirname(self.path) or ".", exist_ok=True)
        with open(self.path, "w") as f:
            json.dump(state, f, ensure_ascii=False, indent=2)

    def get_last_sync(self, project_key: str) -> dict | None:
        return self._load().get(project_key, {}).get("last_sync")

    def get_checkpoint(self, project_key: str) -> dict | None:
        return self._load().get(project_key, {}).get("checkpoint")

    def save_checkpoint(self, project_key: str, started_at: str,
                        items_synced: int, checkpoint_updated_at: str | None):
        state = self._load()
        proj = state.setdefault(project_key, {})
        proj["checkpoint"] = {
            "started_at": started_at,
            "items_synced": items_synced,
            "checkpoint_updated_at": checkpoint_updated_at,
        }
        self._save(state)

    def complete_sync(self, project_key: str, completed_at: str,
                      items_synced: int, sync_type: str):
        state = self._load()
        proj = state.setdefault(project_key, {})
        proj.pop("checkpoint", None)
        proj["last_sync"] = {
            "completed_at": completed_at,
            "items_synced": items_synced,
            "sync_type": sync_type,
        }
        self._save(state)


class SyncService:
    """同期オーケストレーター"""

    def __init__(self, client: JiraClient, db: Database, sync_state: SyncState):
        self.client = client
        self.db = db
        self.sync_state = sync_state

    def build_jql(self, project_key: str, mode: str) -> str:
        """同期モードに応じた JQL を構築"""
        base = f"project = {project_key} ORDER BY updated ASC, key ASC"

        if mode == "resume":
            cp = self.sync_state.get_checkpoint(project_key)
            if cp and cp.get("checkpoint_updated_at"):
                after = cp["checkpoint_updated_at"]
                dt = datetime.fromisoformat(str(after)) - timedelta(minutes=MARGIN_MINUTES)
                date_str = dt.strftime("%Y-%m-%d %H:%M")
                return (f"project = {project_key} "
                        f'AND updated >= "{date_str}" '
                        f"ORDER BY updated ASC, key ASC")

        if mode == "incremental":
            last = self.sync_state.get_last_sync(project_key)
            if last and last.get("completed_at"):
                dt = datetime.fromisoformat(str(last["completed_at"])) - timedelta(minutes=MARGIN_MINUTES)
                date_str = dt.strftime("%Y-%m-%d %H:%M")
                return (f"project = {project_key} "
                        f'AND updated >= "{date_str}" '
                        f"ORDER BY updated ASC, key ASC")

        return base

    def execute(
        self,
        project_key: str,
        mode: str = "full",
        projects: list[dict] | None = None,
        statuses: list[dict] | None = None,
        priorities_df: pd.DataFrame | None = None,
        issue_types_df: pd.DataFrame | None = None,
        fields_df: pd.DataFrame | None = None,
        on_progress: Callable[[int, int], None] | None = None,
    ) -> dict:
        """同期を実行

        Returns:
            {"mode": str, "issues": int, "history": int, "expanded": dict, "summary": dict}
        """
        started_at = datetime.now(timezone.utc).isoformat()
        self.sync_state.save_checkpoint(project_key, started_at, 0, None)

        # 1. Issues 取得
        jql = self.build_jql(project_key, mode)
        logger.info("Sync mode=%s, JQL=%s", mode, jql)
        raw_issues = self.client.fetch_all_issues(jql=jql, on_progress=on_progress)

        # 2. データ変換
        issues_df = IssueTransformer.transform(raw_issues)
        change_history_df = ChangeHistoryTransformer.transform(raw_issues)

        # 3. DB 書き込み
        if projects:
            self.db.upsert_projects(projects)
        self.db.upsert_issues(issues_df, project_key, full=(mode == "full"))

        # チェックポイント更新
        checkpoint_val = None
        if not issues_df.empty:
            checkpoint_val = str(issues_df["updated_date"].max())
            self.sync_state.save_checkpoint(
                project_key, started_at, len(issues_df), checkpoint_val
            )

        self.db.upsert_change_history(change_history_df, project_key)

        if statuses is not None:
            self.db.upsert_metadata(
                project_key, statuses,
                priorities_df if priorities_df is not None else pd.DataFrame(),
                issue_types_df if issue_types_df is not None else pd.DataFrame(),
                fields_df if fields_df is not None else pd.DataFrame(),
            )

        # 4. フィールド展開
        expand_result = {}
        if fields_df is not None and not fields_df.empty:
            expander = FieldExpander(self.db.conn, fields_df)
            expand_result = expander.expand(project_key)

        # 5. 同期完了
        completed_at = datetime.now(timezone.utc).isoformat()
        self.db.record_sync(
            project_key, mode, started_at, completed_at,
            len(issues_df), checkpoint_val,
        )
        self.sync_state.complete_sync(project_key, completed_at, len(issues_df), mode)

        summary = self.db.get_summary()
        logger.info(
            "Sync completed: %d issues, %d history, expanded=%s",
            summary["issues"], summary["history"], expand_result,
        )
        return {
            "mode": mode,
            "fetched": len(issues_df),
            "history": len(change_history_df),
            "expanded": expand_result,
            "summary": summary,
        }

File: jd-go/core/sync.go
package core

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"time"
)

const marginMinutes = 5

// SyncState manages sync state persistence in a JSON file.
type SyncState struct {
	path string
}

// NewSyncState creates a new SyncState.
func NewSyncState(path string) *SyncState {
	return &SyncState{path: path}
}

func (s *SyncState) load() map[string]map[string]json.RawMessage {
	data, err := os.ReadFile(s.path)
	if err != nil {
		return make(map[string]map[string]json.RawMessage)
	}
	var state map[string]map[string]json.RawMessage
	if err := json.Unmarshal(data, &state); err != nil {
		return make(map[string]map[string]json.RawMessage)
	}
	return state
}

func (s *SyncState) save(state map[string]map[string]json.RawMessage) error {
	dir := filepath.Dir(s.path)
	if dir != "" && dir != "." {
		os.MkdirAll(dir, 0o755)
	}
	data, err := json.MarshalIndent(state, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(s.path, data, 0o644)
}

// GetLastSync returns the last sync info for a project.
func (s *SyncState) GetLastSync(projectKey string) (*LastSync, error) {
	state := s.load()
	proj, ok := state[projectKey]
	if !ok {
		return nil, nil
	}
	raw, ok := proj["last_sync"]
	if !ok {
		return nil, nil
	}
	var ls LastSync
	if err := json.Unmarshal(raw, &ls); err != nil {
		return nil, err
	}
	return &ls, nil
}

// GetCheckpoint returns the checkpoint for a project.
func (s *SyncState) GetCheckpoint(projectKey string) (*SyncCheckpoint, error) {
	state := s.load()
	proj, ok := state[projectKey]
	if !ok {
		return nil, nil
	}
	raw, ok := proj["checkpoint"]
	if !ok {
		return nil, nil
	}
	var cp SyncCheckpoint
	if err := json.Unmarshal(raw, &cp); err != nil {
		return nil, err
	}
	return &cp, nil
}

// SaveCheckpoint saves a sync checkpoint.
func (s *SyncState) SaveCheckpoint(projectKey, startedAt string, itemsSynced int, checkpointUpdatedAt *string) error {
	state := s.load()
	if state[projectKey] == nil {
		state[projectKey] = make(map[string]json.RawMessage)
	}
	cp := SyncCheckpoint{
		StartedAt:          startedAt,
		ItemsSynced:        itemsSynced,
		CheckpointUpdatedAt: checkpointUpdatedAt,
	}
	data, _ := json.Marshal(cp)
	state[projectKey]["checkpoint"] = data
	return s.save(state)
}

// CompleteSync marks a sync as complete.
func (s *SyncState) CompleteSync(projectKey, completedAt string, itemsSynced int, syncType string) error {
	state := s.load()
	if state[projectKey] == nil {
		state[projectKey] = make(map[string]json.RawMessage)
	}
	delete(state[projectKey], "checkpoint")
	ls := LastSync{
		CompletedAt: completedAt,
		ItemsSynced: itemsSynced,
		SyncType:    syncType,
	}
	data, _ := json.Marshal(ls)
	state[projectKey]["last_sync"] = data
	return s.save(state)
}

// SyncService orchestrates syncs.
type SyncService struct {
	Client    *JiraClient
	DB        *Database
	SyncState *SyncState
}

// NewSyncService creates a new SyncService.
func NewSyncService(client *JiraClient, db *Database, syncState *SyncState) *SyncService {
	return &SyncService{Client: client, DB: db, SyncState: syncState}
}

// BuildJQL builds a JQL query based on sync mode.
func (svc *SyncService) BuildJQL(projectKey, mode string) string {
	base := fmt.Sprintf("project = %s ORDER BY updated ASC, key ASC", projectKey)

	if mode == "resume" {
		cp, _ := svc.SyncState.GetCheckpoint(projectKey)
		if cp != nil && cp.CheckpointUpdatedAt != nil {
			t, err := time.Parse(time.RFC3339, *cp.CheckpointUpdatedAt)
			if err == nil {
				t = t.Add(-time.Duration(marginMinutes) * time.Minute)
				dateStr := t.Format("2006-01-02 15:04")
				return fmt.Sprintf(`project = %s AND updated >= "%s" ORDER BY updated ASC, key ASC`,
					projectKey, dateStr)
			}
		}
	}

	if mode == "incremental" {
		last, _ := svc.SyncState.GetLastSync(projectKey)
		if last != nil && last.CompletedAt != "" {
			t, err := time.Parse(time.RFC3339, last.CompletedAt)
			if err == nil {
				t = t.Add(-time.Duration(marginMinutes) * time.Minute)
				dateStr := t.Format("2006-01-02 15:04")
				return fmt.Sprintf(`project = %s AND updated >= "%s" ORDER BY updated ASC, key ASC`,
					projectKey, dateStr)
			}
		}
	}

	return base
}

// Execute runs a sync.
func (svc *SyncService) Execute(ctx context.Context, opts SyncOptions) (*SyncResult, error) {
	startedAt := time.Now().UTC().Format(time.RFC3339)
	if err := svc.SyncState.SaveCheckpoint(opts.ProjectKey, startedAt, 0, nil); err != nil {
		slog.Warn("Failed to save initial checkpoint", "error", err)
	}

	// 1. Fetch issues
	jql := svc.BuildJQL(opts.ProjectKey, opts.Mode)
	slog.Info("Sync starting", "mode", opts.Mode, "jql", jql)
	rawIssues, err := svc.Client.FetchAllIssues(ctx, jql, opts.OnProgress)
	if err != nil {
		return nil, fmt.Errorf("fetch issues: %w", err)
	}

	// 2. Transform
	issues := TransformIssues(rawIssues)
	changeHistory := TransformChangeHistory(rawIssues)

	// 3. DB writes
	if len(opts.Projects) > 0 {
		if err := svc.DB.UpsertProjects(opts.Projects); err != nil {
			return nil, fmt.Errorf("upsert projects: %w", err)
		}
	}
	if err := svc.DB.UpsertIssues(issues, opts.ProjectKey, opts.Mode == "full"); err != nil {
		return nil, fmt.Errorf("upsert issues: %w", err)
	}

	// Checkpoint update
	var checkpointVal *string
	if len(issues) > 0 {
		// Find max updated_date
		var maxUpdated string
		for _, iss := range issues {
			if iss.UpdatedDate != nil && *iss.UpdatedDate > maxUpdated {
				maxUpdated = *iss.UpdatedDate
			}
		}
		if maxUpdated != "" {
			checkpointVal = &maxUpdated
			svc.SyncState.SaveCheckpoint(opts.ProjectKey, startedAt, len(issues), checkpointVal)
		}
	}

	if err := svc.DB.UpsertChangeHistory(changeHistory, opts.ProjectKey); err != nil {
		return nil, fmt.Errorf("upsert change history: %w", err)
	}

	if len(opts.Statuses) > 0 || len(opts.Fields) > 0 {
		if err := svc.DB.UpsertMetadata(opts.ProjectKey, opts.Statuses, opts.Priorities, opts.IssueTypes, opts.Fields); err != nil {
			return nil, fmt.Errorf("upsert metadata: %w", err)
		}
	}

	// 4. Field expansion
	var expandResult ExpandResult
	if len(opts.Fields) > 0 {
		expander := NewFieldExpander(svc.DB.DB, opts.Fields)
		var err error
		expandResult, err = expander.Expand(opts.ProjectKey)
		if err != nil {
			slog.Warn("Field expansion failed", "error", err)
		}
	}

	// 5. Complete sync
	completedAt := time.Now().UTC().Format(time.RFC3339)
	svc.DB.RecordSync(opts.ProjectKey, opts.Mode, startedAt, completedAt, len(issues), checkpointVal)
	svc.SyncState.CompleteSync(opts.ProjectKey, completedAt, len(issues), opts.Mode)

	summary, _ := svc.DB.GetSummary()
	slog.Info("Sync completed", "issues", summary.Issues, "history", summary.History, "expanded", expandResult)

	return &SyncResult{
		Mode:     opts.Mode,
		Fetched:  len(issues),
		History:  len(changeHistory),
		Expanded: expandResult,
		Summary:  summary,
	}, nil
}
